torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -0,0 +1,109 @@
+import math
+
+import numpy as np
+import torch
+import torch.distributed as dist
+
+
+def _orthogonalize(matrix, epsilon=1e-8):
+    """Applies Gram-Schmidt procedure to orthogonalize a given 2D tensor."""
+    num_cols = matrix.shape[1]
+    for i in range(num_cols):
+        # Normalize the i'th column.
+        col = matrix[:, i : i + 1]
+        # Note that col ** 2 can underflow/overflow if we use FP16.
+        # May need to consder multiplying a scaling factor and divding it later, or using bfloat16 isntead.
+        col /= torch.sqrt(torch.sum(col ** 2))
+        # If no epsilon is added here, division by zero may be caused by vanishing gradients.
+        # This epsilon is not needed if the input matrix covers the gradients of at least one entire layer in the neural network.
+        if epsilon > 0:
+            col += epsilon
+        # Project it on the rest and remove it.
+        if i + 1 < num_cols:
+            rest = matrix[:, i + 1 :]
+            rest -= torch.sum(col * rest, dim=0) * col
+
+
+def _set_random(vector, random_seed=0):
+    """
+    Initializes a 1D tensor from a standard normal distribution.
+    The see makes sure that the randomized vector in all the DDP replicas are the same at every step.
+    """
+    torch.manual_seed(np.random.RandomState(random_seed).randint(1_000_000_000))
+    vector.data[:] = torch.randn(*vector.shape)
+
+
+def powerSGD_hook(
+    process_group: object, bucket: dist._GradBucket, rank: int = 1
+) -> torch.futures.Future:
+    """
+    This DDP communication hook implements a simplified PowerSGD graidient compression
+    algorithm described in https://arxiv.org/abs/1905.13727.
+    Once gradient tensors are aggregated across all workers, this hook applies
+    compression as follows:
+    1) Views the input flattened 1D gradient tensor as a square-shaped tensor M with 0 paddings;
```

Comment:
I'm curious if this works. This would yield even stronger compression than regular rank-1/2 PowerSGD, where there are more smaller matrices that are not necessarily square.

Reply:
The input tensor in my hook is actually a flattened 1D tensor that batches multiple per-variable tensors, corresponding to multiple neighboring layers in the network.

Here the input tensor does not need to have a square shape, and 0 paddings will be added to make a square.

The problem I found is that if I directly apply the original alg, if there are too many small tensors, the computation costs of orthogonizing and matrix multiplications over small matrices can be expensive. Here is more like a batching.